src/net/socket_address.rs

```diff
@@ -0,0 +1,81 @@
+#![allow(unsafe_code)]
+use core::mem::size_of;
+
+use crate::backend::{
+    c,
+    net::write_sockaddr::{encode_sockaddr_v4, encode_sockaddr_v6},
+};
+
+use super::{SocketAddr, SocketAddrV4, SocketAddrV6};
+
+#[cfg(unix)]
+use super::SocketAddrUnix;
+
+/// Opaque type equivalent to `sockaddr` in C.
+///
+/// This is always used behind a raw pointer that is cast from a pointer to a
+/// `sockaddr`-compatible C type, and then cast back to a `sockaddr` pointer
+/// to be passed to a system call.
+#[repr(C)]
+pub struct SockAddrRaw {
+    _data: [u8; 0],
+}
```

Comment:
What would you think of renaming `SockAddrRaw` to `SocketAddrOpaque`?

There are two changes there:
 - `Sock` -> `Socket`, because if we have some things prefixed `Socket` and some prefixed `Sock` I expect I'll frequently forget which things use which prefixes.
 - `Raw -> `Opaque`, which I think better captures that this struct type doesn't describe the data layout.